core/base/v7/inc/ROOT/TDirectory.h

```diff
@@ -148,35 +148,31 @@ class TDirectory {
   /// \throws TDirectoryTypeMismatch if the object stored under this name is of
   ///   a type that is not a derived type of `T`.
   template <class T>
-  std::shared_ptr<T> Get(const std::string& name) {
-    ContentMap_t::iterator idx = fContent.find(name);
+  decltype(auto) Get(const std::string& name) {
+    auto idx = fContent.find(name);
     if (idx != fContent.end()) {
-      // FIXME: implement upcast!
-      if (auto dep
-        = dynamic_cast<Internal::TDirectoryEntryPtr<T>*>(idx->second.get())) {
-        return dep->GetPointer();
-      }
+      if (idx->second.GetTypeInfo() == typeid(ToContentType_t<T>))
+        return std::static_pointer_cast<ToContentType_t<T>>(idx->second.GetPointer());
+      if (auto ptr = idx->second.CastPointer<ToContentType_t<T>>())
+        return ptr;
       // FIXME: add expected versus actual type name as c'tor args
       throw TDirectoryTypeMismatch(name);
     }
     throw TDirectoryUnknownKey(name);
-    return std::shared_ptr<T>(); // never happens
+    return std::shared_ptr<ToContentType_t<T>>(); // never happens
```

Comment:
Isn't this the same as `return {nullptr}`? I think I prefer that; it's simpler, more obvious.